### DISCUSSION TITLE
Sharing my results - feedback welcome :)

### DISCUSSION THREAD

**Discussion by IvonaTau** [JCBrouwer/maua-stylegan2#8] [Show and tell]
First of all, great work! I really enjoyed playing with this repo :) 

Here's a video I created using a stylegan2 trained on some of my architecture images:
https://vimeo.com/492821043

(for comparison there is also a "vanilla" interpolation of the same model: https://vimeo.com/492822184)  

*  Due to perspective in the training images the reflective pad seems to create warping in the buildings - is there a quick fix to remove it? 

* I was also wondering is it possible to increase the number of latents used in the video (I have seen that the default number (12) is dependent on the number of notes in chromagram - so perhaps not so easy to change)

**Comment by JCBrouwer** [comment]
Huh, seems like something isn't going right with the timing, the video is longer than the audio. Did you change the fps or load in your own audio files (not via audio_file flag)?

I actually kind of like the weird warping effect, but it does definitely detract from the realism. The reflective padding is done in get_bends() to make the output 1920x1080. You can remove it by commenting out `get_bends=get_bends` in `generate()`. Although you'll then also have to set the `out_size=1024`.

Another thing you could try is doubling the width a bit higher up in the network (more pixels = smoother transition). e.g. at the 32x32 layer you could duplicate each column of the feature map and blur it. Hopefully that would create a perspective warp that's a bit smoother/monotone over the course of the image (although it might also still seem stretched).

Something like this could do the trick (I haven't actually tested this code...)
```
import kornia.torchgeometry.image as ktgmi

def get_bends(args):
    class DoubleColumns(th.nn.Module):
        def forward(self, x):
            new_x = []
            for i in range(x.shape[-1]):        # for each coord in the last dimension
                new_x += [x[..., [i]], x[..., [i]]] # add two copies of the column to the list
            return th.cat(new_x, axis=-1)   # concatenate the list along the last dimension

    tfm = th.nn.Sequential([
        DoubleColumns(),                        # 1x3x32x32 --> 1x3x32x64
        ktgmi.GaussianBlur((3, 3), (1, 1)) # blur it a little for even smoother output
    ])

    # apply transform to 7th layer (which is 32x32 in default 1024 stylegan2 network)
    return [{"layer": 5, "transform": tfm}]
```
Do note that your feature maps will only be double the width after layer 5. When you set `out_size=1920` the noise width and height fed to `get_noise()` will be double width, so you'll have to add a case to that function to make sure the noise maps generated for the first layers are square (not yet doubled). You can just add:
```
...
if scale < 5: # the same number as layer from the transform above
    width /= 2
...
```

To use more latents in your video, you can take a look at [`slerp_loops`](https://github.com/JCBrouwer/maua-stylegan2/blob/master/audioreactive/latent.py#L48) or [`spline_loops`](https://github.com/JCBrouwer/maua-stylegan2/blob/master/audioreactive/latent.py#L83). These functions take a list of latents and return an interpolations of n_frames long looping n_loops time over the course of it. They use different strategies to loop, so you can just try them both out and see which you like more.

Another option is to split your audio out into stems using [spleeter](https://github.com/deezer/spleeter) and blending multiple `chroma_weight_latents()` generated from different stems.

**Reply by JCBrouwer** [reply]
> Huh, seems like something isn't going right with the timing, the video is longer than the audio. Did you change the fps or load in your own audio files (not via audio_file flag)?

I've figured out the issue! If your audio is less than 60 seconds long (the default duration I've set just so people don't try to render 20 min on their first try), then the video gets stretched out to that length.

You can fix it by just removing the duration=60 line from `generate()`. Although, I've also pushed a fix so cloning the most recent version of the repo shouldn't see this behavior anymore!

@IvonaTau
